### PR TITLE
util: add xval

### PR DESCRIPTION
**Pull request by BusyJay**
@ngaut @coocood @qiuyesuifeng PTAL

### CODE REVIEW

File: src/util/xeval/evaluator.rs [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+        Ok(cmp.map(|c| c > Ordering::Equal).into())
+    }
+
+    fn eval_null_eq(&self, expr: &Expr) -> Result<Datum> {
+        let (left, right) = try!(self.eval_children(expr));
+        let cmp = try!(left.cmp(&right));
+        Ok((cmp == Ordering::Equal).into())
+    }
+
+    fn cmp_children(&self, expr: &Expr) -> Result<Option<Ordering>> {
```
Comment (coocood) [inline src/util/xeval/evaluator.rs]: cmp_two_chilrdren

File: src/util/xeval/evaluator.rs [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+        let (left, right) = try!(self.eval_children(expr));
+        if left == Datum::Null || right == Datum::Null {
+            return Ok(None);
+        }
+        left.cmp(&right).map(Some).map_err(From::from)
+    }
+
+    fn eval_children(&self, expr: &Expr) -> Result<(Datum, Datum)> {
```
Comment (coocood) [inline src/util/xeval/evaluator.rs]: eval_two_children

File: src/util/xeval/evaluator.rs [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+        self.eval_children_as_bool(expr)
+            .map(|(l, r)| (l.unwrap_or(false) && r.unwrap_or(false)).into())
+    }
+
+    fn eval_or(&self, expr: &Expr) -> Result<Datum> {
+        self.eval_children_as_bool(expr)
+            .map(|(l, r)| (l.unwrap_or(false) || r.unwrap_or(false)).into())
+    }
+
+    fn eval_children_as_bool(&self, expr: &Expr) -> Result<(Option<bool>, Option<bool>)> {
```
Comment (coocood) [inline src/util/xeval/evaluator.rs]: eval_two_children_as_bool

**Comment by ngaut** [comment]
PTAL @shenli

File: src/util/xeval/evaluator.rs [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+            ExprType::LE => self.eval_le(expr),
+            ExprType::EQ => self.eval_eq(expr),
+            ExprType::NE => self.eval_ne(expr),
+            ExprType::GE => self.eval_ge(expr),
+            ExprType::GT => self.eval_gt(expr),
+            ExprType::NullEQ => self.eval_null_eq(expr),
+            ExprType::And => self.eval_and(expr),
+            ExprType::Or => self.eval_or(expr),
+            ExprType::Float32 | ExprType::Float64 => unimplemented!(),
+            _ => Ok(Datum::Null),
```
Comment (ngaut) [inline src/util/xeval/evaluator.rs]: Is it possible we run to this branch ? @coocood

**Comment by ngaut** [comment]
Rest LGTM.

File: src/util/xeval/evaluator.rs (hunk 4 above)
Comment (qiuyesuifeng) [inline src/util/xeval/evaluator.rs]: Should we return Err？

Comment (shenli) [inline src/util/xeval/evaluator.rs]: @ngaut I don't think so. TiKV coprocessor can not scan table or index now. Maybe in the next pr.

**Comment by ngaut** [comment]
LGTM

**Comment by qiuyesuifeng** [comment]
LGTM

**Pull request merged**